examples/cvae/cvae.py

```diff
@@ -0,0 +1,188 @@
+        # are fed into the prior network.
+        self.baseline_net = pre_trained_baseline_net
+        self.prior_net = Encoder(z_dim, hidden_1, hidden_2)
+        self.generation_net = Decoder(z_dim, hidden_1, hidden_2)
+        self.recognition_net = Encoder(z_dim, hidden_1, hidden_2)
+
+    def model(self, xs, ys=None):
+        # register this pytorch module and all of its sub-modules with pyro
+        pyro.module("generation_net", self)
+        batch_size = xs.shape[0]
+        with pyro.plate("data"):
+
+            # Prior network uses the baseline predictions as initial guess.
+            # This is the generative process with recurrent connection
+            with torch.no_grad():
+                # this ensures the training process does not change the
+                # baseline network
+                y_hat = self.baseline_net(xs).view(xs.shape)
+
+            # sample the handwriting style from the prior distribution, which is
+            # modulated by the input xs.
+            prior_loc, prior_scale = self.prior_net(xs, y_hat)
+            zs = pyro.sample('z', dist.Normal(prior_loc, prior_scale).to_event(1))
+
+            # the output y is generated from the distribution pθ(y|x, z)
+            loc = self.generation_net(zs)
+
+            if ys is not None:
+                # In training, we will only sample in the masked image
+                mask_loc = loc[(xs == -1).view(-1, 784)].view(batch_size, -1)
+                mask_ys = ys[xs == -1].view(batch_size, -1)
+                pyro.sample('y', dist.Bernoulli(mask_loc).to_event(1), obs=mask_ys)
+            else:
+                # In testing, no need to sample: the output is already a
+                # probability in [0, 1] range, which better represent pixel
+                # values considering grayscale. If we sample, we will force
+                # each pixel to be  either 0 or 1, killing the grayscale
+                pyro.deterministic('y', loc.detach())
+
+            # return the loc so we can visualize it later
+            return loc
+
+    def guide(self, xs, ys=None):
+        with pyro.plate("data"):
+            if ys is None:
+                # at inference time, ys is not provided. In that case,
+                # the model uses the prior network
+                y_hat = self.baseline_net(xs).view(xs.shape)
+                loc, scale = self.prior_net(xs, y_hat)
+            else:
+                # at training time, uses the variational distribution
+                # q(z|x,y) = normal(loc(x,y),scale(x,y))
+                loc, scale = self.recognition_net(xs, ys)
+
+            pyro.sample("z", dist.Normal(loc, scale).to_event(1))
+
+    def save(self, model_path):
+        torch.save({
+            'prior': self.prior_net.state_dict(),
+            'generation': self.generation_net.state_dict(),
+            'recognition': self.recognition_net.state_dict()
+        }, model_path)
+
+    def load(self, model_path, map_location=None):
```

Comment:
@fritzo there's probably a better way to do this? use `PyroModule`?

Reply:
I think `PyroModule` would be overkill for this model. However @ucals is it possible to simplify `save`/`load` as follows:
```py
def save(self, model_path):
    torch.save(self, model_path)

def load(self, model_path, map_location=None):
    loaded = torch.load(model_path, map_location=map_location)
    self.load_state_dict(loaded.state_dict())
    self.eval()
```
And if that works, I don't see why you really need custom `.save()`, `.load()` methods at all; user code could simply
```py
torch.save(model_path, cvae)
...
cvae = torch.load(model_path)
cvae.eval()
```

Reply:
You're right! These custom methods were completely unnecessary, just tested! Thanks for the tip!